src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs

```diff
@@ -268,7 +268,13 @@ private void GenerateCheckBox(ModelExplorer modelExplorer, TagHelperOutput outpu
                 if (hiddenForCheckboxTag != null)
                 {
                     hiddenForCheckboxTag.TagRenderMode = renderingMode;
-                    output.Content.Append(hiddenForCheckboxTag);
+
+                    var hiddenForCheckboxList = ViewContext.FormContext.FormData.ContainsKey("hiddenForCheckBoxList") ?
+                        ViewContext.FormContext.FormData["hiddenForCheckBoxList"] as List<TagBuilder> :
```

Comment:
Either this cast should be unconditional or you need a `Debug.Assert()` before using the value on line 276.

Also don't do the lookup twice: Use `TryGetValue()`.